src/contract.ts

```diff
@@ -1,376 +1,479 @@
+     */
+    async load(program: Signer, so: Buffer, payer?: Signer | null): Promise<void> {
+        if (!program.publicKey.equals(this.program)) throw new Error('INVALID_PROGRAM_ACCOUNT'); // @FIXME: add error types
+
+        payer ||= this.payer;
+        if (!payer) throw new Error('MISSING_PAYER_ACCOUNT'); // @FIXME: add error types
+
+        // @TODO: error if the program already exists without sending a transaction
+
+        await BpfLoader.load(this.connection, payer, program, so, BPF_LOADER_PROGRAM_ID);
+    }
+
+    /**
+     * Deploy the contract to a loaded Solana program.
+     *
+     * @param name            Name of the contract to deploy
+     * @param constructorArgs Arguments to pass to the contract's Solidity constructor function
+     * @param program         Keypair for the account the program is located at
+     * @param storage         Keypair for the account the program's data is stored at
+     * @param space           Byte size to allocate for the storage account (this cannot be resized)
+     * @param options         @TODO: docs
+     *
+     * @return @TODO: docs
+     */
+    async deploy(
+        name: string,
+        constructorArgs: any[],
+        program: Signer,
+        storage: Signer,
+        space: number,
+        options?: ContractCallOptions
+    ): Promise<ContractCallResult> {
+        if (!program.publicKey.equals(this.program)) throw new Error('INVALID_PROGRAM_ACCOUNT'); // @FIXME: add error types
+        if (!storage.publicKey.equals(this.storage)) throw new Error('INVALID_STORAGE_ACCOUNT'); // @FIXME: add error types
+
+        const payer = options?.payer || this.payer;
+        if (!payer) throw new Error('MISSING_PAYER_ACCOUNT'); // @FIXME: add error types
+
+        const {
+            accounts = [],
+            writableAccounts = [],
+            programDerivedAddresses = [],
+            signers = [],
+            sender = payer.publicKey,
+            value = 0,
+            simulate = false,
+            confirmOptions = {
+                commitment: 'confirmed',
+                skipPreflight: false,
+                preflightCommitment: 'processed',
+            },
+        } = options ?? {};
+
+        const hash = keccak256(Buffer.from(name));
+        const seeds = programDerivedAddresses.map((pda) => pda.seed);
+        const input = this.interface.encodeDeploy(constructorArgs);
+
+        const data = Buffer.concat([
+            this.storage.toBuffer(), //                     storage @FIXME: these comments are kind of useless
+            sender.toBuffer(), //                           sender  @FIXME: better to explain why, not what
+            Buffer.from(numToPaddedHex(value), 'hex'), //   value
+            Buffer.from(hash.substr(2, 8), 'hex'), //       hash
+            encodeSeeds(seeds), //                          seeds
+            Buffer.from(input.replace('0x', ''), 'hex'), // input
+        ]);
```

Comment:
For these comments, I was hoping for a high level description of the serialization format and why it expects data this way.

Reply:
I can improve this once this is merged.